src/main/java/com/puppycrawl/tools/checkstyle/ant/CheckstyleAntTask.java

```diff
@@ -509,33 +524,114 @@ private Properties createOverridingProperties() {
 
     /**
      * Returns the list of files (full path name) to process.
-     * @return the list of files included via the filesets.
+     * @return the list of files included via the filesets and paths.
      */
-    protected List<File> scanFileSets() {
-        final List<File> list = new ArrayList<>();
+    protected List<File> getFilesToCheck() {
+        final List<File> allFiles = new ArrayList<>();
         if (fileName != null) {
             // oops we've got an additional one to process, don't
             // forget it. No sweat, it's fully resolved via the setter.
             log("Adding standalone file for audit", Project.MSG_VERBOSE);
-            list.add(new File(fileName));
+            allFiles.add(new File(fileName));
+        }
+
+        final List<File> filesFromFileSets = scanFileSets();
+        allFiles.addAll(filesFromFileSets);
+
+        final List<File> filesFromPaths = scanPaths();
+        allFiles.addAll(filesFromPaths);
+
+        return allFiles;
+    }
+
+    /**
+     * Retrieves all files from the defined paths.
+     * @return a list of files defined via paths.
+     */
+    protected List<File> scanPaths() {
+        final List<File> allFiles = new ArrayList<>();
+
+        for (int i = 0; i < paths.size(); i++) {
+            final Path currentPath = paths.get(i);
+            final List<File> pathFiles = scanPath(currentPath, i + 1);
+            allFiles.addAll(pathFiles);
+        }
+
+        return allFiles;
+    }
+
+    /**
+     * Scans the given path and retrieves all files for the given path.
+     *
+     * @param path      A path to scan.
+     * @param pathIndex The index of the given path. Used in log messages only.
+     * @return A list of files, extracted from the given path.
+     */
+    private List<File> scanPath(Path path, int pathIndex) {
+        final String[] resources = path.list();
+        log(pathIndex + ") Scanning path " + path, Project.MSG_VERBOSE);
+        final List<File> allFiles = new ArrayList<>();
+        final List<File> concreteFiles = new ArrayList<>();
```

Comment:
@rnveach 
> Why don't you remove concreteFiles, add everything to allFiles, and use a primitive int variable to count what you add under if (file.isFile()) { for logging.

This will be a right decision. There is no need in allocating the memory for that list,